[MLIR][Linalg] pack, unpack to take memref inputs

ita9naiwa opened the pull request:
#129004 

- Change to use `ShapedType`, not `RankedTensorType`
  - accordingly add branches to handle both `MemrefType` and `TensorType`
  - Add Memref::CollapseShapeOp::inferCollapsedType to handle some `MemrefType`

ita9naiwa committed:
draft

ita9naiwa committed:
draft

ita9naiwa committed:
draft

ita9naiwa committed:
init

ita9naiwa committed:
lint

ita9naiwa committed:
lint

ita9naiwa committed:
add

ita9naiwa committed:
remove tensor casting

ita9naiwa committed:
add test

adam-smnk commented:
I expect most of the existing `linalg.pack/unpack` transforms can't handle memrefs. It'd be great to restrict them to ops with pure tensor semantics.

hanhanW reviewed (changes requested):
@ita9naiwa thanks for your patch! I think we can break the PR into a couple of PRs:

- Add the support for pack/unpack on memrefs. In this PR, you update the td file and apply those `ShpedType` changes. For the transformations, we bail out if the op does not have pure tensor semantics, i.e., check with `hasPureTensorSemantics`. And add a roundtrip test: https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Linalg/roundtrip.mlir
- Add the transformations support for memrefs. I don't have a use case for lowering on memrefs, but we can contribute to it if it is helpful.
- Add the generateScalarImplementation support + a test in `loops.mlir`. I don't expect to see a change in `loops.mlir` with this change.

ita9naiwa commented:
Thanks @hanhanW ! I wasn't sure how to proceed but now it became clear!

hanhanW commented:
> > Thanks @hanhanW ! I wasn't sure how to proceed but now it became clear!
> 
> Out of curiosity, what’s the use case for lowering to memref? I’m okay with moving forward with these changes, but I’d prefer to have a clear use case since extending to memref would require additional work (e.g., bufferization).

I think it is more for completeness. IMO, all the linalg ops should support both tensor and memref. There is an [issue](https://github.com/llvm/llvm-project/issues/129004) for tracking it. In this context, we need an e2e reasonable fallback for the lowering, and all the linalg ops implement `generateScalarImplementation`. Since we are moving pack/unpack to linalg dialect, we have to fill the gaps between the new ops and linalg ops.

IMO, the few gaps are:
- Missing support for memref version
- bufferization
- generateScalarImplementation

Bufferization should be easy by the design of linalg ops. If we can't do it through linalg interface way, we also have [the implementation in IREE](https://github.com/iree-org/iree/blob/6ecafc10149a0ee5fa7bcba171dd3774211b7577/compiler/src/iree/compiler/Codegen/Interfaces/BufferizationInterfaces.cpp#L361-L499) that can be upstreamed.

chelini commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -5002,9 +5001,12 @@ static bool isLikePadUnPad(PackOrUnpackOp packOp,
+  if (auto packedTensorType =
+          llvm::dyn_cast<RankedTensorType>((*this)->getResultTypes().front()))
+    return isLikePadUnPad(*this, packedTensorType);
+  if (auto packedTensorType =
```
why this if? We can cast directly to ShapedType.

chelini commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -2525,6 +2526,35 @@ MemRefType CollapseShapeOp::computeCollapsedType(
 }
 
+MemRefType
+CollapseShapeOp::inferCollapsedType(MemRefType type,
```
this code is really similar to this https://github.com/llvm/llvm-project/blob/d38380d3d808183652c1e9be34e3a2476ed6ea70/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp#L1816 Can we share more code between the two methods?

ita9naiwa replied:
Is it okay to make template function inside `mlir/include/mlir/IR/TypeUtilities.h`?

ita9naiwa replied:
```c++
template <typename ShapedTy>
ShapedTy inferCollapsedType(ShapedTy type, ArrayRef<AffineMap> reassociation) {
  static_assert(std::is_same<ShapedTy, RankedTensorType>::value ||
                std::is_same<ShapedTy, MemRefType>::value,
                "Expected RankedTensorType or MemRefType");

  auto shape = type.getShape();
  SmallVector<int64_t, 4> newShape;
  newShape.reserve(reassociation.size());

  assert(isReassociationValid(reassociation) && "invalid reassociation");
  unsigned currentDim = 0;
  for (AffineMap m : reassociation) {
    unsigned dim = m.getNumResults();
    auto band = shape.slice(currentDim, dim);
    int64_t size = 1;
    if (llvm::is_contained(band, ShapedType::kDynamic))
      size = ShapedType::kDynamic;
    else
      for (unsigned d = 0; d < dim; ++d)
        size *= shape[currentDim + d];
    newShape.push_back(size);
    currentDim += dim;
  }

  return ShapedTy::get(newShape, type.getElementType());
}
```

ita9naiwa replied:
Sorry, we don't need `CollapseShapeOp::inferCollapsedType` for MemrefType at all.

Commit:
Merge branch 'main' into ita9naiwa/pack-memref

ita9naiwa commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -1799,6 +1799,11 @@ def MemRef_CollapseShapeOp : MemRef_ReassociativeReshapeOp<"collapse_shape", [
 
     static MemRefType computeCollapsedType(
         MemRefType srcType, ArrayRef<ReassociationIndices> reassociation);
+    static MemRefType
```
We don't need them, I removed and changed to use computeCollapsedType instead.

Commit:
fix upon review

ita9naiwa commented on mlir/include/mlir/Dialect/Linalg/IR/RelayoutOpInterface.td:
```diff
@@ -10,6 +10,7 @@
 #define LINALG_IR_RELAYOUTOPINTERFACE
 
 include "mlir/Interfaces/DestinationStyleOpInterface.td"
+include "mlir/Dialect/Linalg/IR/LinalgInterfaces.td"
```
Removed!

ita9naiwa commented:
> Should be a Variadic<AnyRankedTensor> - generally linalg ops don't have return values at memref (no aliasing)

I'm implementing separating cases for
1. Tensor Semantic -> return $result
2. Memref semantic -> fill $dest

when it's done, I'll re-request reviews.

Commit:
lint

Commit:
Merge branch 'main' into ita9naiwa/pack-memref

banach-space commented:
Hi @ita9naiwa 👋🏻 

IIUC, this is upstreaming code from IREE? If yes, would you mind listing the original authors as co-authors in the summary? For reference, LLVM's notes on code-review process when upstreaming code:
* https://llvm.org/docs/CodeReview.html#upstreaming-changes-to-llvm

Thank you 🙏🏻

ita9naiwa commented:
I'll surely add! thanks for mentioning it! @banach-space

Commit:
format fix

ita9naiwa commented:
This isn't upstreaming from IREE.

hanhanW commented on mlir/test/Dialect/Linalg/roundtrip.mlir:
```diff
@@ -706,3 +706,21 @@ func.func @conv2d_channel_first_q_promote(%img: tensor<100x3x224x224xi8>, %filt:
 // CHECK:         linalg.conv_2d_nchw_fchw_q {dilations = dense<1> : tensor<2xi64>, strides = dense<1> : tensor<2xi64>} ins(%[[arg0]], %[[arg1]], %[[arg2]], %[[arg3]] : tensor<100x3x224x224xi8>, tensor<64x3x5x5xi8>, i8, i8) outs(%{{.*}} : tensor<100x64x220x220xi32>) -> tensor<100x64x220x220xi32>
+
+// -----
+// Test that we can lower all the way to LLVM without crashing, don't check results here.
```
I don't follow the comment.. Can you elaborate?

ita9naiwa replied:
Just wanted to check if mlir can lower to llvm ir.

hanhanW replied:
But this is a roundtrip test. It verifies that the `mlir-opt` can parse the IR and print the IR that matches the check. There are no LLVM IR involved.

ita9naiwa replied:
I'll check an appropriate file to put this test later; now I can't find one.

hanhanW replied:
What I meant is that it is legacy. Removing the comment is suitable to me. No need to check that

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -396,18 +397,30 @@ linalg::lowerUnPack(RewriterBase &rewriter, linalg::UnPackOp unPackOp,
-      RankedTensorType::Builder(packedTensorType).setShape(stripMinedShape);
-  RankedTensorType collapsedType = tensor::CollapseShapeOp::inferCollapsedType(
-      stripMinedTensorType, packingMetadata.reassociations);
+  ShapedType stripMinedType;
+  if (auto tensorType = packedTensorType.dyn_cast<TensorType>()) {
+    stripMinedType =
+        RankedTensorType::get(stripMinedShape, tensorType.getElementType());
+  } else if (auto memrefType = packedTensorType.dyn_cast<MemRefType>()) {
+    stripMinedType =
+        MemRefType::get(stripMinedShape, memrefType.getElementType());
+  }
+  ShapedType collapsedType;
+  if (stripMinedType.isa<TensorType>()) {
+    collapsedType = tensor::CollapseShapeOp::inferCollapsedType(
+        cast<RankedTensorType>(stripMinedType), packingMetadata.reassociations);
+  } else if (stripMinedType.isa<MemRefType>()) {
+    collapsedType = memref::CollapseShapeOp::computeCollapsedType(
+        cast<MemRefType>(stripMinedType), packingMetadata.reassociations);
+  }
```
I don't expect the change in the PR, and I think we agree to move the transformation to a separate PR right?

ita9naiwa replied:
Sorry, I'll revert this change.

hanhanW commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
Please revert the changes from the file because it is not part of the PR.

ita9naiwa replied:
Done

hanhanW commented on mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:
Why do we need the change in the PR?

ita9naiwa replied:
Done

hanhanW commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -315,11 +315,11 @@ SmallVector<Range> SliceFromCollapseHelper::getExtractSliceParams(
-          llvm::map_range(it.value(), [&](int64_t idx) -> Range {
-            return {zeroAttr, collapseShapeInputShape[idx], oneAttr};
-          }));
+      llvm::append_range(offsetsSizesAndStrides,
+                         llvm::map_range(it.value(), [&](int64_t idx) -> Range {
+                           return {zeroAttr, collapseShapeInputShape[idx],
+                                   oneAttr};
+                         }));
```
I don't know what happens to the lint, but do we really need the change for the PR?

ita9naiwa replied:
Done

hanhanW commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -43,10 +43,10 @@ class Linalg_RelayoutOp<string mnemonic, list<Trait> traits = []> :
-      return ::llvm::cast<RankedTensorType>(getSource().getType()); };
-    RankedTensorType getDestType() {
-      return ::llvm::cast<RankedTensorType>(getDest().getType()); };
+    ShapedType getSourceType() {
```
In the above traits, we need to remove the `NoMemoryEffect` trait and implement the trait like other LinalgOps. I.e., we need `DeclareOpInterfaceMethods<MemoryEffectsOpInterface>` for pack/unpack ops. Here is the [implmentation](https://github.com/iree-org/iree/blob/ce2585aa5db62151157e9af4b0449109e9335f4c/compiler/src/iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp#L58-L78) in IREE. We should have something similar in `LinalgOps.cpp`.

Side note: other linalg ops have similar implementation using the [getGenericEffectsImpl method](https://github.com/llvm/llvm-project/blob/8f3f93cd78cfbf1dea349be2eef98802da8ad929/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L1212-L1236)

If we don't have the trait, I think the memref version would always be DCE-ed. E.g., if you run `mlir-opt --cse repro.mlir`, the pack op is not preserved, which is wrong.


```mlir
func.func @pack_memref(%source: memref<128x256xf32>) -> memref<8x16x8x32xf32> {
  %dest = memref.alloc() : memref<8x16x8x32xf32>
  linalg.pack %source outer_dims_perm = [1, 0] inner_dims_pos = [0, 1] inner_tiles = [8, 32]
      into %dest : memref<128x256xf32> -> memref<8x16x8x32xf32>
  return %dest : memref<8x16x8x32xf32>
}
```

becomes

```mlir
module {
  func.func @pack_memref(%arg0: memref<128x256xf32>) -> memref<8x16x8x32xf32> {
    %alloc = memref.alloc() : memref<8x16x8x32xf32>
    return %alloc : memref<8x16x8x32xf32>
  }
}
```

ita9naiwa replied:
Done!

ita9naiwa replied:
checked, and now it's not DCEed

hanhanW commented on mlir/test/Dialect/Linalg/roundtrip.mlir:
```diff
@@ -706,3 +706,21 @@ func.func @conv2d_channel_first_q_promote(%img: tensor<100x3x224x224xi8>, %filt:
+
+// -----
+// Test that we can lower all the way to LLVM without crashing, don't check results here.
+func.func @pack_memref(%source: memref<128x256xf32>) -> memref<8x16x8x32xf32> {
+  %dest = memref.alloc() : memref<8x16x8x32xf32>
```
nit: pass `dest` as a function argument.

ita9naiwa replied:
Done

Commit:
revert changes

Commit:
revert changes

Commit:
nit

hanhanW commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -76,6 +76,13 @@ class Linalg_RelayoutOp<string mnemonic, list<Trait> traits = []> :
     /// with `inner_dims_pos` rather than the packed tensor.
     SmallVector<int64_t> getTiledOuterDims();
+
+         void $cppClass::getEffects(
+         SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+             &effects) {
+       getGenericEffectsImpl(effects, cast<LinalgOp>(getOperation()));
+     }
```
I think it is better to put the implementation to LinalgOps.cpp. And I believe that this is not working for relayout ops because they are not LinalgOp. I think you need to implement it separately and you can follow what we've done in IREE.

https://github.com/llvm/llvm-project/blob/6075275e68bd2362095af76b1acb49c2ac9610a4/mlir/include/mlir/Dialect/Linalg/IR/LinalgInterfaces.td#L225-L229

ita9naiwa replied:
I implemented getEffects, but their implementations are identical for packOp, unpackOp, I think it would be great to wrap somewhere, could you suggest some?

hanhanW replied:
Can we do it with template? E.g., they have the identical implementation for `reifyResultShapes`, and we have the implementation in `reifyResultShapesImpl`. I think we can use the same trick here.

https://github.com/llvm/llvm-project/blob/8244f8210f2e62f68429a0daf104fd483ada45ab/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L4330-L4341

ita9naiwa replied:
Thanks! Done!

Commit:
fix upon review: Add getEffects for PackOp and UnPackOp

ita9naiwa commented:
I removed as much transformation/vectorization code, but two places it was necessary to add some `dyn_cast` for compilation. @hanhanW

Commit:
make clang-format happy

Commit:
make clang-format happy

Commit:
wrap getEffects function

Commit:
Merge branch 'main' into ita9naiwa/pack-memref

hanhanW commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -190,7 +190,7 @@ def Linalg_PackOp : Linalg_RelayoutOp<"pack", [
     // tiles, position of the inner tiles (innerDimsPos)  and interchange vector
     // of outer loops (outerDimsPerm).
-    static RankedTensorType inferPackedType(RankedTensorType sourceType,
+    static RankedTensorType inferPackedType(ShapedType sourceType,
```
This looks more like a `inferPackedTensorType` to me now. Because it always returns a `RankedTensorType`. In the LLVM codebase, it is used in op verification for shapes and the data-layout propagation pass. IMO, we can leave it as what it is and update the verifier.

We can add a new `inferShape` to the pack op and use it in the verifier.

https://github.com/llvm/llvm-project/blob/4c4e4e4299b16d8dd85811e9dd8697b17c95577f/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L4471-L4480

On the other hand, please revisit the verifier. I'd expect some changes there. E.g., we do not support unranked tensor/memref in Linalg ops, at least for these pack/unpack ops. Without your change, it emits an error like `custom op 'linalg.pack' invalid kind of type specified`. With your change, it crashes. This is because we switch to `AnyShaped`, and there is no longer "hasRank" trait in the op definition.

To repro: `mlir-opt repro.mlir`

```mlir
func.func @pack_tensor(%source: tensor<*xf32>, %dest: tensor<*xf32>) -> tensor<*xf32>{
  %0 = linalg.pack %source outer_dims_perm = [] inner_dims_pos = [] inner_tiles = []
      into %dest : tensor<*xf32> -> tensor<*xf32>                                                                                                                                                                                                              return %0 : tensor<*xf32>
}
```

IMO, we should check that the types are ranked, and the shapes are compatible.

ita9naiwa replied:
```cpp
// Verify that the source and destination are ranked types.
  if (!packOrUnPack.getSourceType().hasRank() ||
      !packOrUnPack.getDestType().hasRank()) {
    return op->emitError("expected both source and destination to be shaped types");
  }
 ```
  
I think we can put this,

hanhanW reviewed (commented):
Thanks, here is the other round of review comments. I think we need to make some changes in verifier as well.

ita9naiwa commented on mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:
```diff
@@ -573,7 +573,7 @@ PackingMetadata computePackingMetadata(int64_t packedRank,
 /// shape. When an optional cst attribute is passed, it is reshaped only if the
 /// splat value matches the value in the attribute.
-OpFoldResult reshapeConstantSource(DenseElementsAttr source, TensorType result,
+OpFoldResult reshapeConstantSource(DenseElementsAttr source, ShapedType result,
```
Can we modify like this?
```
OpFoldResult PackOp::fold(FoldAdaptor adaptor) {
  if (!hasPureTensorSemantics())
    return {};

  std::optional<Attribute> paddingValue;
  if (auto pad = adaptor.getPaddingValue())
    paddingValue = pad;
  if (OpFoldResult reshapedSource = reshapeConstantSource(
          llvm::dyn_cast_if_present<DenseElementsAttr>(adaptor.getSource()),
          cast<TensorType>(getDestType()), paddingValue))
    return reshapedSource;
  return {};
}
```

Not sure if I understand your comment correctly.

ita9naiwa commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -4801,6 +4801,45 @@ PackOp PackOp::createTransposedClone(OpBuilder &b, Location loc,
 }
 
+template <typename OpTy>
+static void getEffectsImpl(
```
Done!

ita9naiwa commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -4951,7 +4993,7 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
       packOp.getSourceMutable().assign(source);
       packOp.getDestMutable().assign(dest);
-      packOp.getResult().setType(cast<RankedTensorType>(dest.getType()));
+      packOp.getResult().setType(cast<ShapedType>(dest.getType()));
```
now tensor version, memref version works well.

I think it's good idea to put these into test cases; how do you think?

```mlir
module {
func.func @fold_pack_unpack_memref(%x: memref<2x3xf32>) -> memref<2x3xf32> {
  %unpacked = linalg.unpack %x outer_dims_perm = [] inner_dims_pos = [] inner_tiles = []
               into %x : memref<2x3xf32> -> memref<2x3xf32>
  %packed = linalg.pack %unpacked outer_dims_perm = [] inner_dims_pos = [] inner_tiles = []
             into %x : memref<2x3xf32> -> memref<2x3xf32>
  return %packed : memref<2x3xf32>
}
}
```
is canonicalized into
```mlir
module {
  func.func @fold_pack_unpack_memref(%arg0: memref<2x3xf32>) -> memref<2x3xf32> {
    %unpack = linalg.unpack %arg0 inner_dims_pos = [] inner_tiles = [] into %arg0 : memref<2x3xf32> -> memref<2x3xf32>
    return %arg0 : memref<2x3xf32>
  }
}
```


```mlir
module {
func.func @fold_pack_unpack_tensor(%x: tensor<2x3xf32>) -> tensor<2x3xf32> {
  %unpacked = linalg.unpack %x outer_dims_perm = [] inner_dims_pos = [] inner_tiles = []
             into %x : tensor<2x3xf32> -> tensor<2x3xf32>
  %packed = linalg.pack %unpacked outer_dims_perm = [] inner_dims_pos = [] inner_tiles = []
             into %x : tensor<2x3xf32> -> tensor<2x3xf32>
  return %packed : tensor<2x3xf32>
}
}
```
reduces into
```mlir
base ❯ mlir-opt --canonicalize --cse cano-tensor.mlir
module {
  func.func @fold_pack_unpack_tensor(%arg0: tensor<2x3xf32>) -> tensor<2x3xf32> {
    return %arg0 : tensor<2x3xf32>
  }
}
```

hanhanW replied:
SGTM, please add the tests to https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Linalg/canonicalize.mlir

EDIT: You should have inner_dims_pos in the `fold_pack_unpack_memref` test. (We can have canonicalization patterns to fold them away if the configuration is empty and the types statically match.)

ita9naiwa commented on mlir/lib/Dialect/Linalg/Transforms/PackAndUnpackPatterns.cpp:
Surely I believe to do so. is there a good way to verify patterns related to tensor pack/unpack ops?

Commit:
fix upon review

ita9naiwa commented on mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h:
```diff
@@ -451,7 +451,7 @@ getLinearizedDimensions(ArrayRef<ReassociationIndices> reassociationIndices);
 ///          tensor<3x7x11x10xf32> to tensor<1x1x1x10xf32>
 ///
-///    %5 = tensor.collapse_shape %4 [[0, 1, 2], [3]] : 
+///    %5 = tensor.collapse_shape %4 [[0, 1, 2], [3]] :
```
This happens every time I run `clang-format`, this is not relevant, I'll remove this change when final commit before merge. is it fine?

ita9naiwa commented:
Hi @hanhanW , I mostly addressed your review comments. canonicalization pattern works for Tensor and memref.

I want to know how to track down all the patterns related to Tensor Pack/UnPack Ops

Commit:
bail out transforms using PackOp, UnPackOp

ita9naiwa commented:
> I want to know how to track down all the patterns related to Tensor Pack/UnPack Ops

I tracked down with grep matching `linalg::PackOp` and `linalg::UnPackOp`.

I bailed out transformations and rewrite patterns using e.g., 

```C++
// TODO(issues/129004): Support MemRef PackOp. Temporarily return failure.
  if (!packOp.hasPureTensorSemantics()) {
    return failure();
  }
```

Commit:
fix build error

Commit:
fix build error

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -4930,23 +5002,31 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
       auto newDestType = packOp.getDestType().clone(destShape);
-      dest =
-          rewriter.create<tensor::CastOp>(loc, newDestType, packOp.getDest());
+      if (hasTensorSemantics)
+        dest =
+            rewriter.create<tensor::CastOp>(loc, newDestType, packOp.getDest());
```
There are three types in the pack ops on tensors. (1) source type (2) dest type (3) result type.

In the shape inference, we need casting for (1) and (2), so here you also need to take memref into account. (A new test will capture the failure). For (3), where is updated in the `modifyOpInPlace{...}`, we update the result type if and only if it is on tensors.

The (3) only happens on tensors because memref variant only has (1) and (2) types.

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -4956,9 +5036,15 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
-      auto castOp =
-          rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
-      rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
+      if (hasTensorSemantics) {
+        auto castOp =
+            rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
+        rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
+      } else {
+        auto castOp =
+            rewriter.create<memref::CastOp>(loc, originalResultType, packOp);
+        rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
+      }
```
nit: we can declare `Operation* castOp` and move the rewrite out from the if-else body.

ita9naiwa replied:
It should be done like
```MLIR
  Operation* castOp;
  if (hasTensorSemantics) {
    castOp = rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
  } else {
    castOp = rewriter.create<memref::CastOp>(loc, originalResultType, packOp);
  }
  rewriter.replaceAllUsesExcept(packOp, castOp->getResult(0), castOp);
```

is it correct?

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -5038,6 +5126,9 @@ struct FoldTensorCastPackOp : public OpRewritePattern<PackOp> {
     if (!tensor::hasFoldableTensorCastOperand(op))
       return failure();
 
+    if (!op.hasPureTensorSemantics())
```
Please add a TODO for memref. It is valid to fold memref.cast away as well.

ita9naiwa replied:
Done

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/BlockPackMatmul.cpp:
```diff
@@ -91,6 +91,11 @@ transposePackedMatmul(RewriterBase &rewriter, linalg::LinalgOp linalgOp,
                       linalg::PackOp packOp, AffineMap operandMap,
                       ArrayRef<unsigned> blocksStartDimPos,
                       bool transposeOuterBlocks, bool transposeInnerBlocks) {
+  // TODO(issues/129004): Support MemRef PackOp. Temporarily return failure.
+  if (!packOp.hasPureTensorSemantics()) {
+    return failure();
+  }
```
Style nit 1: do not reference issue ID in TODO, I think it is LLVM style.
Style nit 2: do not use braces for the simple single-statement: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:
```diff
@@ -63,6 +63,13 @@ getPackingInfoFromOperand(OpOperand *opOperand, linalg::GenericOp genericOp,
   static_assert(llvm::is_one_of<OpTy, linalg::PackOp, linalg::UnPackOp>::value,
                 "applies to only pack or unpack operations");
+  // TODO(issues/129004): Support MemRef PackOp. Temporarily return failure.
+  if (auto linalgOp =
+          dyn_cast<linalg::LinalgOp>(packOrUnPackOp.getOperation())) {
```
The condition will always be false because pack/unpack does not implement LinalgOp interface. I think it is unclear to do layout propagation on memrefs because there are memory scope ish things. So maybe drop the TODO and just return failure when it does not have pure tensor semantics.

EDIT: we don't need the change in the local function because we should already bail out if they are on memrefs. This is dead code.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -355,11 +360,18 @@ FailureOr<LowerPackResult> linalg::lowerPack(RewriterBase &rewriter,
   rewriter.setInsertionPoint(unPackOp);
 
-  RankedTensorType packedTensorType = unPackOp.getSourceType();
+  // TODO: support non-ranked tensor types. ShapedType
+  RankedTensorType packedTensorType =
+      dyn_cast<RankedTensorType>(unPackOp.getSourceType());
```
We are not going to support non-ranked types for pack/unpack ops. So please remove the TODO.

style nit: dyn_cast already spells the type, please use auto here. https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
Some checks are redundant. I think we only add the checks to precondition functions.

hanhanW commented on mlir/test/Dialect/Linalg/roundtrip.mlir:
```diff
@@ -706,3 +706,17 @@ func.func @conv2d_channel_first_q_promote(%img: tensor<100x3x224x224xi8>, %filt:
 // CHECK:   %[[arg0:[a-zA-z0-9]*]]: tensor<100x3x224x224xi8>, %[[arg1:[a-zA-z0-9]*]]: tensor<64x3x5x5xi8>, %[[arg2:[a-zA-z0-9]*]]: i8, %[[arg3:[a-zA-z0-9]*]]: i8)
 // CHECK:         linalg.conv_2d_nchw_fchw_q {dilations = dense<1> : tensor<2xi64>, strides = dense<1> : tensor<2xi64>} ins(%[[arg0]], %[[arg1]], %[[arg2]], %[[arg3]] : tensor<100x3x224x224xi8>, tensor<64x3x5x5xi8>, i8, i8) outs(%{{.*}} : tensor<100x64x220x220xi32>) -> tensor<100x64x220x220xi32>
+
+// -----
+func.func @pack_memref(%source: memref<128x256xf32>, %dest: memref<8x16x8x32xf32>) {
+  linalg.pack %source outer_dims_perm = [1, 0] inner_dims_pos = [0, 1] inner_tiles = [8, 32]
+      into %dest : memref<128x256xf32> -> memref<8x16x8x32xf32>
+  return
+}
+
+// -----
+func.func @unpack_memref(%source: memref<16x8x8x32xf32>, %dest: memref<128x256xf32>) {
+  linalg.unpack %source inner_dims_pos = [0, 1] inner_tiles = [8, 32]
+      into %dest : memref<16x8x8x32xf32> -> memref<128x256xf32>
+  return
+}
```
Let's add a blank line right after the separator (i.e., `// -----`), and please add checks. Here is the doc for FileCheck: https://llvm.org/docs/CommandGuide/FileCheck.html

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp:
I think we are not going to support layout propagation on memref, so we can delete all the TODO from the file.

hanhanW reviewed (changes requested):
Looks better to me, thanks! RE TODO comments: I only expect that they are available for PackAndUnpackPatterns, Vectorization, and canonicalization. Other transform might not able to handle the memref case atm. I suggest removing the TODO from other files, and people can start their projects later if they need it.

Commit:
Merge branch 'main' into ita9naiwa/pack-memref

Commit:
add invalid pack/unpack cases

Commit:
fix roundtrip test

Commit:
fix upon review

Commit:
fix upon review

Commit:
.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1664,12 +1668,17 @@ static LogicalResult
   rewriter.setInsertionPoint(unpackOp);
 
-  RankedTensorType unpackTensorType = unpackOp.getSourceType();
+  auto unpackTensorType = dyn_cast<RankedTensorType>(unpackOp.getSourceType());
+  if (!unpackTensorType)
+    return failure();
```
ditto: can be simplified with `cast` and remove the check.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1889,6 +1898,10 @@ vectorizeDynamicLinalgOpPrecondition(linalg::LinalgOp op,
 vectorizeUnPackOpPrecondition(linalg::UnPackOp unpackOp,
                               ArrayRef<int64_t> inputVectorSizes) {
+  // TODO: Support Memref PackOp. Temporarily return failure.
+  if (!unpackOp.hasPureTensorSemantics()) {
+    return failure();
+  }
```
style nit: remove braces

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
I think I have a similar comment before. Can you remove the checks from `vectorizeAsTensorPackOp` and `vectorizeAsTensorUnpackOp`? How we set up vectorization is to check everything in `*Precondition()`, so other checks are redundant to me.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
```diff
@@ -1030,6 +1042,10 @@ static Value getPackOpSourceOrPaddedSource(OpBuilder &builder,
     return input;
   }
 
+  // TODO: Support Memref PackOp. Temporarily return failure.
+  if (!packOp.hasPureTensorSemantics())
+    return packOp.getSource();
```
The code does not agree with the comment. How about returning `Value()` when the op does not have pure tensor semantics? Please also update the function comment to reflect the change.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Transforms.cpp:
Please remove all the TODOs from this file. I think they are not trivial, because we don't have `memref.pad` op. No one will clear the TODO in this case.

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -9,8 +9,8 @@
 //===----------------------------------------------------------------------===//
-
 #include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include <iostream>
```
IWYU, please delete the include.

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -5330,6 +5423,9 @@ struct FoldTensorCastUnPackOp : public OpRewritePattern<UnPackOp> {
     if (!tensor::hasFoldableTensorCastOperand(op))
       return failure();
 
+    if (!op.hasPureTensorSemantics())
+      return failure();
```
Let's add a TODO for consistency. It is a reasonable folder to me and we should support it (in follow-ups).

hanhanW reviewed (commented):
Thank you for addressing most of the comments. I think we are close to land the PR. I left some comments about TODO and comments, and I think we are missing some tests for canonicalization patterns. Please add tests to reflect the changes.

Commit:
Merge branch 'main' into ita9naiwa/pack-memref

ita9naiwa committed:
fix upon review

Signed-off-by: Hyunsung Lee <ita9naiwa@gmail.com>

ita9naiwa commented:
I accidently  clang-formatted invalid.mlir, I will fix very soon.

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -4942,7 +5015,7 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
     }
     Value dest = packOp.getDest();
-    RankedTensorType originalResultType = packOp.getDestType();
+    ShapedType originalResultType = packOp.getDestType();
```
This canonicalization pattern is broken when pack/unpack are on memrefs. We should only update the result type when they are on tensors. The memref version does not return any SSA value.

```
    rewriter.modifyOpInPlace(packOp, [&] {
      packOp.getSourceMutable().assign(source);
      packOp.getDestMutable().assign(dest);
      packOp.getResult().setType(cast<RankedTensorType>(dest.getType()));
    });
```

Also, the above cast should be memref.cast if we are going to support it in the PR. If you don't know how to implement it, we can bail out in line 5009 and I can help add the support later.

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -4957,9 +5030,19 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
     if (needUpdateDestType) {
       rewriter.setInsertionPointAfter(packOp);
-      auto castOp =
-          rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
-      rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
+      Operation *castOp;
+      bool hasTensorSemantics = packOp.hasPureTensorSemantics();
+      if (hasTensorSemantics) {
+        castOp =
+            rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
+      } else {
+        castOp =
+            rewriter.create<memref::CastOp>(loc, originalResultType, packOp);
+      }
+      rewriter.replaceAllUsesExcept(packOp, castOp->getResult(0), castOp);
+    } else {
+      // TODO: support memref.cast if static shape inference is available.
+      return failure();
```
This does not make sense to me. This casting is needed only for tensor version. I was wrong in the previous review.

Reason: the memref does not have return value at all. The casting is needed for making type consistent. E.g.,

```mlir
%pack = linalg.pack %src into %dest ... -> tensor<?x?xf32>
return %pack : tensor<?x?xf32> // some uses
```

after the shape inference, the result type is changed in the rewriter:

```

%pack = linalg.pack %casted_src into %casted_dest ... -> tensor<3x4xf32>
// This is where the types mismatch, so we need a final cast for this.
return %pack : tensor<?x?xf32>
```

It is not the case for memref because memref does not have any return value at all.

So you only need to update the if condition, which becomes

```cpp
if (needUpdateDestType && packOp.hasPureTensorSemantics()) {
  rewriter.setInsertionPointAfter(packOp);
  auto castOp =
      rewriter.create<tensor::CastOp>(loc, originalResultType, packOp);
   rewriter.replaceAllUsesExcept(packOp, castOp, castOp);
}
```

hanhanW commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -190,7 +190,30 @@ def Linalg_PackOp : Linalg_RelayoutOp<"pack", [
     // tiles, position of the inner tiles (innerDimsPos)  and interchange vector
     // of outer loops (outerDimsPerm).
-    static RankedTensorType inferPackedType(RankedTensorType sourceType,
+    /// This method uses inferPackedShape to ensure consistency with other shape
+    /// inference methods regarding which dimensions are dynamic.
```
This is implementation details which should not be documented here. I think we can just remove it..

hanhanW commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -190,7 +190,30 @@ def Linalg_PackOp : Linalg_RelayoutOp<"pack", [
+    // Method to get the `MemRefType` of the result based on the inner
+    // tiles, position of the inner tiles (innerDimsPos)  and interchange vector
+    // of outer loops (outerDimsPerm).
+    /// This method uses inferPackedShape to ensure consistency with other shape
+    /// inference methods regarding which dimensions are dynamic.
```
same here

hanhanW commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td:
```diff
@@ -190,7 +190,30 @@ def Linalg_PackOp : Linalg_RelayoutOp<"pack", [
+        ArrayRef<int64_t> innerTileSizes, ArrayRef<int64_t> innerDimsPos,
+        ArrayRef<int64_t> outerDimsPerm = {});
+
+    // Method to get the Shape of the result based on the input shape, inner
+    // tiles, position of the inner tiles (innerDimsPos)  and interchange vector
+    // of outer loops (outerDimsPerm).
+
+    /// Helper for PackOp::{getResultShape, inferPackedTensorType, inferPackedMemRefType}.
+    /// Returns the shape of the packed type. Having a shared helper helps
+    /// implement these three methods in a way that ensures
+    /// that they agree on which dimensions are dynamic.
```
```suggestion
    // Returns the shape of the packed type. It is a shared helper helps type inference methods in a way that ensures  that they agree on which dimensions are dynamic.
```

hanhanW reviewed (commented):
The canonicalization pattern is off, which leads to an invalid IR. Other parts look good, just some nits about comments.

ita9naiwa committed:
Update mlir/include/mlir/Dialect/Linalg/IR/LinalgRelayoutOps.td

Co-authored-by: Han-Chung Wang <hanhan0912@gmail.com>
Signed-off-by: Hyunsung Lee <ita9naiwa@gmail.com>

hanhanW reviewed (commented):
Sorry that I made a mistake in the review about the folding. After taking a look at the real IR tests, I think there are two issues.

1. The memref version should not return a memref. By definition, the op performs packing from the source buffer and store the result to the destination buffer. Like other linalg operations, it should not return any value when it has buffer semantics. In other linalg ops, the `Variadic<AnyRankedTensor>` adds the check. It was pointed out in [a comment](https://github.com/llvm/llvm-project/pull/129036/files#r1977098388) from @adam-smnk, and I think it is not resolved.

Invalid case:

```mlir
%packed = linalg.pack %unpacked inner_dims_pos = [0, 1] inner_tiles = [4, 4] into %buf_packed : memref<40x80xf32> -> memref<10x20x4x4xf32>
```

It should be like this valid case:

```mlir
linalg.pack %unpacked
  inner_dims_pos = [0, 1] inner_tiles = [4, 4]
  into %buf_packed
  : memref<40x80xf32> -> memref<10x20x4x4xf32>
```

2. The second issue is about the folding on memrefs. TLDR is that we should disable it. Because the current logic is not correct. How I'd implement the folding is that we need to look at the source memref and other uses of the source memref. It could be expensive and the logic is more complicated on memrefs.  The below is not foldable if there are other ops using `%buf_unpacked` in the middle. Things become way more complicated when control flow is involved. I think this kind of folding should be implemented in a pass manner. (Please remember to remove the test cases in `canonicalzation.mlir` if they are disabled. I also did not see the reason why you added some folding tests to tensor, while the PR is scoped to add the support for memref. Maybe we can drop the tests for tensors as well?)

```mlir
linalg.unpack %t
  inner_dims_pos = [0, 1] inner_tiles = [4, 4]
  into %buf_unpacked
  : memref<10x20x4x4xf32> -> memref<40x80xf32>
linalg.pack %unpacked
  inner_dims_pos = [0, 1] inner_tiles = [4, 4]
  into %buf_packed
  : memref<40x80xf32> -> memref<10x20x4x4xf32>
```

(The rest issue is about reverting non-related code changes, I think you mentioned that you will remove them before landing the PR.)

hanhanW commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -4747,13 +4770,21 @@ SmallVector<OpFoldResult> PackOp::getResultShape(
 /// Get the expected packed type based on source type, tile factors, position of
 /// the inner tiles and permutation of the outer tiled loop.
-RankedTensorType PackOp::inferPackedType(RankedTensorType sourceType,
+RankedTensorType PackOp::inferPackedTensorType(
```
bump

ita9naiwa commented:
Apologies for the delay — I’ve been recovering from a medical issue. I’ll resume this soon.

ita9naiwa committed:
Revert unnecessary cosmetic changes

- Revert OpenMP conversion changes
- Revert DataLayoutPropagation tensor semantic checks removal
- Revert Vectorization tensor semantic checks removal
- Revert Transforms tensor semantic checks removal
- Revert ReshapeOpsUtils formatting changes
- Revert MemRefOps.td whitespace changes

Keep only essential changes for memref support in pack/unpack

ita9naiwa committed:
Address HanHan review feedback: disable canonicalization for memref pack/unpack

- Add hasPureTensorSemantics() check at the beginning of PackOp::canonicalize()
- Add hasPureTensorSemantics() check at the beginning of UnPackOp::canonicalize()
- Remove memref folding tests from canonicalize.mlir
- Add tests to verify memref pack/unpack canonicalization is disabled

This prevents complex canonicalization patterns from running on memref
versions of pack/unpack operations, following buffer semantics.

ita9naiwa committed:
Merge branch 'main' into ita9naiwa/pack-memref

ita9naiwa committed:
Apply clang-format to code changes

ita9naiwa committed:
fix upon review
